src/vsphere_cpi/lib/cloud/vsphere/vcenter_client.rb

```diff
@@ -217,54 +217,69 @@ def find_network_retryably(datacenter, network_name)
-          valid_networks = network_container.portgroup
-        elsif network_container.instance_of?(VimSdk::Vim::Folder)
-          valid_networks = network_container.child_entity
-        end
+    def find_network(datacenter, name)
+      # Split the name into an optional container part and name part
+      *container_path, name = name.split('/')
+
+      # Find the container if specified
+      if !container_path.empty?
+        container = find_by_inventory_path([datacenter.name, 'network', *container_path])
+        container ||= find_child_by_name(datacenter.mob.network_folder, container_path)
+        raise NetworkNotFoundError, "Can't find network container #{container_path.join('/')}" if container.nil?
       else
-        valid_networks = datacenter.mob.network
+        container = nil
       end
 
-      target_network = nil
-      matching_networks = valid_networks.select do |network|
-        begin
-          network.name == network_name
-        rescue => e
-          logger.warn("Can't retrieve name of network #{network}: #{e}")
-          false
-        end
-      end
-      if matching_networks.length == 1
-        target_network = matching_networks.first
-      elsif matching_networks.length > 1
-        # pick the Standard Portgroup if multiple networks exist with the given name
-        target_network = matching_networks.find { |n| n.instance_of?(VimSdk::Vim::Network) }
-        if target_network.nil?
-          raise "Multiple networks found for #{network_name}. Please specify the full path, for example 'FOLDER_NAME/DISTRIBUTED_SWITCH_NAME/DISTRIBUTED_PORTGROUP_NAME'"
-        end
+      # Make the list of candidate networks from the container
+      networks = case container
+        when VimSdk::Vim::Dvs::VmwareDistributedVirtualSwitch
+          container.portgroup
+        when VimSdk::Vim::Folder
+          container.child_entity
+        when nil
+          datacenter.mob.network
       end
 
-      target_network
+      # Find networks that match the network name
+      networks.select! do |network|
+        network.name == name
+      rescue => e
+        logger.warn("Can't retrieve name of network #{network}: #{e}")
+        false
+      end
+
+      return nil if networks.empty?
+      return networks.first if networks.length == 1
+      # If it is VC 7.0, then NSXT 3.0 can configure same Logical switch as separate dvpgs
+      # under each VC cluster's VDS. But all such dvpg map back to same LS in NSXT.
+      # Hence, disambiguation is needed
+      # Check if all portgroups are actually same by checking these conditions
+      # 1. Network type is DVPG
+      # 2. Network responds to backing type (or check vc 7.0) and Network has nsx backing type
+      # 3. Logical Switch ID for DVPG exists && it is same for all (Same as first newtwork in list)
+      network = networks.first if networks.all? do |n|
+        n.is_a?(VimSdk::Vim::Dvs::DistributedVirtualPortgroup) &&
```

Comment:
Yeah I think turning this into a series of `#select`s followed by a single `#all` will make it easier to read.

Reply:
We cannot select and then apply all?  because all of these networks must adhere to conditions. If there is a network that is not meeting conditions, we do not want to deselect and ignore but we want all? to return false.

Reply:
Oh you're totally right.